### PR TITLE
[Snippets] SoftmaxDecomposition is moved to data flow pipeline

### PR DESCRIPTION
**Pull request by v-Golubev**
### Details:
 - *Control flow pass `SoftmaxDecomposition` removed*
 - *Introduced data flow `SoftmaxDecomposition` pass which decomposes `Softmax` operation to a sequence of `Eltwise` ops + `Reduce` ops*
 - *Introduced control flow `ReduceDecomposition` pass which decomposes Reduce operation to a range of low-level operations on linear IR. Limitation: only reduction by last dimension is supported*

### Tickets:
 - *CVS-112847*

### Prerequisites:
- #21303 

PR between branches for review (all the review discussions are there): [Softmax decomposition moved to data flow pipeline](https://github.com/v-Golubev/openvino/pull/24#top)

### CODE REVIEW

File: src/common/snippets/src/op/subgraph.cpp [hunk 1]
```diff
@@ -402,7 +404,11 @@ void Subgraph::data_flow_transformations(const BlockedShapeVector& blocked_input
         manager.register_pass<snippets::pass::MatMulToBrgemm>();
         manager.register_pass<snippets::pass::FuseTransposeBrgemm>();
         manager.register_pass<snippets::pass::TransposeDecomposition>();
-        manager.register_pass<snippets::pass::SetSoftmaxPorts>();
+        if (getenv("DISABLE_DATA_FLOW_DECOMPOSITION")) {
+            manager.register_pass<snippets::pass::SetSoftmaxPorts>();
+        } else {
+            manager.register_pass<snippets::pass::SoftmaxDecomposition>();
+        }
```
Comment (v-Golubev) [inline src/common/snippets/src/op/subgraph.cpp:411]: Remove debug environment + remove SetSoftmaxPorts transformation

File: src/common/snippets/src/op/subgraph.cpp [hunk 2]
```diff
@@ -433,7 +439,9 @@ void Subgraph::control_flow_transformations(lowered::LinearIR& linear_ir,
 
     lowered::pass::PassPipeline pipeline(lowered_pass_config);
     pipeline.register_pass<lowered::pass::MarkLoops>(vector_size);
+    // TODO: remove SoftmaxDecomposition pass
     pipeline.register_pass<lowered::pass::SoftmaxDecomposition>(vector_size);
```
Comment (v-Golubev) [inline src/common/snippets/src/op/subgraph.cpp:443]: TODO:

1. Remove SoftmaxDecomposition
2. Update unit tests which are used this pass
3. Remove a workaround in PropagatePrecisions

File: src/plugins/intel_cpu/src/emitters/x64/cpu_generator.cpp [hunk 3]
```diff
@@ -52,6 +52,15 @@ namespace ov {
...
 }
 
+#define CREATE_UNDEFINED_EMITTER(node_type) { \
+    [](const snippets::lowered::ExpressionPtr& expr) -> std::shared_ptr<snippets::Emitter> { \
+        return nullptr; \
+    }, \
+    [](const std::shared_ptr<ov::Node>& n) -> std::set<std::vector<element::Type>> { \
+        return node_type::get_supported_precisions(n); \
+    } \
+}
```
Comment (v-Golubev) [inline src/plugins/intel_cpu/src/emitters/x64/cpu_generator.cpp:62]: Change the macros: it shouldn't use static method of a node: target machine can report supported precisions for such operations

File: src/plugins/intel_cpu/src/emitters/x64/cpu_generator.cpp [hunk 4]
```diff
@@ -165,6 +174,8 @@ intel_cpu::CPUTargetMachine::CPUTargetMachine(dnnl::impl::cpu::x64::cpu_isa_t ho
     jitters[snippets::op::LoopEnd::get_type_info_static()] = CREATE_SNIPPETS_EMITTER(LoopEndEmitter);
     jitters[intel_cpu::BrgemmCPU::get_type_info_static()] = CREATE_SNIPPETS_EMITTER(BrgemmEmitter);
     jitters[intel_cpu::BrgemmCopyB::get_type_info_static()] = CREATE_SNIPPETS_EMITTER(BrgemmCopyBEmitter);
+    jitters[snippets::op::ReduceMax::get_type_info_static()] = CREATE_UNDEFINED_EMITTER(snippets::op::ReduceMax);
+    jitters[snippets::op::ReduceSum::get_type_info_static()] = CREATE_UNDEFINED_EMITTER(snippets::op::ReduceSum);
```
Comment (v-Golubev) [inline src/plugins/intel_cpu/src/emitters/x64/cpu_generator.cpp:178]: Use Reduce from openvino opset

**Review by a-sidorova — APPROVED** [review]
👍🏼 

My comments were applied in the https://github.com/v-Golubev/openvino/pull/24.
However, I have some [concerns](https://github.com/v-Golubev/openvino/pull/24#discussion_r1467230290) for now regarding default precisions for cases when a node doesn't have an emitter.

**Review by IvanNovoselov — APPROVED** [review]

File: src/common/snippets/include/snippets/lowered/pass/reduce_decomposition.hpp [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+namespace lowered {
+namespace pass {
+
+/**
+ * @interface ReduceDecomposition
+ * @brief Decomposes snippets::Reduce operations to a range of low-level operations on linear IR
+ * @attention Only Reduce by last dimension is supported
+ * @ingroup snippets
+ */
+class ReduceDecomposition : public RangedPass {
```
Comment (dmitry-gorokhov) [inline src/common/snippets/include/snippets/lowered/pass/reduce_decomposition.hpp:20]: As we discussed offline: the follow-up task is to move `ReduceDecomposition` on data-flow stage.

**Review by dmitry-gorokhov — APPROVED** [review]

**Comment by dmitry-gorokhov** [comment]
@v-Golubev BTW, I think we need to update that list https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_cpu/src/extension.cpp#L144

File: src/common/snippets/include/snippets/lowered/pass/reduce_decomposition.hpp (hunk 5 above)
Comment (v-Golubev) [reply src/common/snippets/include/snippets/lowered/pass/reduce_decomposition.hpp:20]: The follow-up ticket is created: 134837

**Comment by v-Golubev** [comment]
> @v-Golubev BTW, I think we need to update that list https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_cpu/src/extension.cpp#L144

You are right, thanks for the finding. Done in [7d2bf26](https://github.com/openvinotoolkit/openvino/pull/21672/commits/7d2bf26edd576822a115503b90052bd5e7886a9a)

**Pull request merged**